File: pippip-native-ios/Messages/ConversationViewController.m
//
//  ConversationViewController.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 2/11/18.
//  Copyright © 2018 seComm. All rights reserved.
//

#import "ConversationViewController.h"
#import "ConversationDataSource.h"
#import "Conversation.h"
#import "ApplicationSingleton.h"
#import "MessageManager.h"
#import "ContactDatabase.h"
#import "AlertErrorDelegate.h"

@interface ConversationViewController ()
{
    MessageManager *messageManager;
    ConversationDataSource *dataSource;
}

@property (weak, nonatomic) IBOutlet UITextField *messageText;
@property (weak, nonatomic) IBOutlet UITableView *conversationTableView;
@property (weak, nonatomic) IBOutlet UIStackView *stackView;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *stackViewBottom;
@property (weak, nonatomic) IBOutlet UILabel *sendFailedLabel;
@property (weak, nonatomic) IBOutlet UINavigationItem *navItem;

@property (weak, nonatomic) ConversationCache *conversationCache;

@end

@implementation ConversationViewController

@synthesize errorDelegate;

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    dataSource = [[ConversationDataSource alloc] initWithTableView:_conversationTableView withPublicId:_publicId];
    _conversationTableView.dataSource = dataSource;
    [_conversationTableView setDelegate:dataSource];
    [_sendFailedLabel setHidden:YES];

    messageManager = [[MessageManager alloc] init];
    [messageManager setResponseConsumer:self];
    // Public ID is set in the messages table prepare for segue.
    _conversationCache = [ApplicationSingleton instance].conversationCache;

    ContactDatabase *contactDatabase = [[ContactDatabase alloc] init];
    NSDictionary *contact = [contactDatabase getContact:_publicId];
    NSString *nickname = contact[@"nickname"];
    if (nickname != nil) {
        _navItem.title = nickname;
    }
    else {
        NSString *fragment = [_publicId substringWithRange:NSMakeRange(0, 6)];
        _navItem.title = [fragment stringByAppendingString:@"..."];
    }

    errorDelegate = [[AlertErrorDelegate alloc] initWithViewController:self withTitle:@"Messages Error"];

}

- (void)viewDidAppear:(BOOL)animated {

    [[NSNotificationCenter defaultCenter] postNotificationName:@"ConversationLoaded" object:nil];

}

- (void)viewWillAppear:(BOOL)animated {

    Conversation *conversation = [_conversationCache getConversation:_publicId];

    if ([conversation count] > 0) {
        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:[conversation count] - 1 inSection:0];
        [_conversationTableView scrollToRowAtIndexPath:indexPath
                                      atScrollPosition:UITableViewScrollPositionBottom
                                              animated:YES];
    }

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardWillShow:)
                                                 name:UIKeyboardWillShowNotification
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardDidHide:)
                                                 name:UIKeyboardDidHideNotification
                                               object:nil];

}

- (void)viewWillDisappear:(BOOL)animated {

    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIKeyboardDidHideNotification object:nil];

}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)keyboardWillShow:(NSNotification*)notify {

    // get height of visible keyboard
    NSDictionary* keyboardInfo = [notify userInfo];
    NSValue* keyboardFrame = [keyboardInfo valueForKey:UIKeyboardFrameEndUserInfoKey];
    CGRect keyboardFrameRect = [keyboardFrame CGRectValue];
    CGFloat keyboardHeight = keyboardFrameRect.size.height;
    _stackViewBottom.constant = keyboardHeight + 2;
    [_stackView setNeedsDisplay];
    Conversation *conversation = [_conversationCache getConversation:_publicId];
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:[conversation count] - 1 inSection:0];
    [_conversationTableView scrollToRowAtIndexPath:indexPath
                                  atScrollPosition:UITableViewScrollPositionBottom
                                          animated:YES];

}

- (void)keyboardDidHide:(NSNotification*)notify {

    _stackViewBottom.constant = 5;

}

- (void)response:(NSDictionary *)info {

    BOOL success = NO;
    if (info != nil) {
        NSString *result = info[@"result"];
        if ([result isEqualToString:@"sent"]) {
            success = YES;
            NSNumber *sq = info[@"sequence"];
            NSNumber *ts = info[@"timestamp"];
            NSString *publicId = info[@"publicId"];
            [messageManager messageSent:publicId
                           withSequence:[sq integerValue]
                          withTimestamp:[ts integerValue]];

            NSMutableDictionary *messageCount = [NSMutableDictionary dictionary];
            messageCount[@"count"] = [NSNumber numberWithUnsignedInteger:1];
            [[NSNotificationCenter defaultCenter] postNotificationName:@"MessagesUpdated" object:nil userInfo:messageCount];

            dispatch_async(dispatch_get_main_queue(), ^{
                [_sendFailedLabel setHidden:success];
                if (success) {
                    _messageText.text = @"";
                    Conversation *conversation = [_conversationCache getConversation:_publicId];
                    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:[conversation count] - 1 inSection:0];
                    [_conversationTableView scrollToRowAtIndexPath:indexPath
                                                  atScrollPosition:UITableViewScrollPositionBottom
                                                          animated:YES];
                }
            });
        }
    }

}

- (IBAction)clearAllMessages:(UIBarButtonItem *)sender {

    [_conversationCache deleteAllMessages:_publicId];
    [dataSource messagesCleared];
    [_conversationTableView reloadData];

}

- (IBAction)sendMessage:(id)sender {

    if (_messageText.text.length > 0) {
        [_sendFailedLabel setHidden:YES];
        NSString *messageText = _messageText.text;
        [messageManager sendMessage:messageText withPublicId:_publicId];
    }

}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: pippip-native-ios/Messages/ConversationViewController.swift
//
//  ConversationViewController.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 4/4/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit
import ChameleonFramework
import RKDropdownAlert

class ConversationViewController: AsyncMessagesViewController, RKDropdownAlertDelegate {

    var conversationDataSource: ConversationDataSource?
    var conversationDelegate: ConversationCollectionDelegate
    var deferredDataSource: DeferredDataSource
    @objc var contact: Contact?
    var selectView: SelectContactView?
    var messageManager = MessageManager()
    var suspended = false
    var returnFromSuspend = false
    var authView: AuthViewController?

    init?() {

        conversationDelegate = ConversationCollectionDelegate()
        deferredDataSource = DeferredDataSource()
        
        super.init(dataSource: deferredDataSource,delegate: conversationDelegate)

    }
    
    required init(coder aDecoder: NSCoder) {

        conversationDelegate = ConversationCollectionDelegate()
        deferredDataSource = DeferredDataSource()

        super.init(coder: aDecoder)

    }

    override init?(tableViewStyle style: UITableViewStyle) {

        conversationDelegate = ConversationCollectionDelegate()
        deferredDataSource = DeferredDataSource()
        
        super.init(dataSource: deferredDataSource,delegate: conversationDelegate)

    }

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        let storyboard = UIStoryboard(name: "Main", bundle: nil)
        authView =
            storyboard.instantiateViewController(withIdentifier: "AuthViewController") as? AuthViewController
        
        NotificationCenter.default.addObserver(self, selector: #selector(contactSelected(_:)),
                                               name: Notifications.ContactSelected, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(messageAdded(_:)),
                                               name: Notifications.MessageAdded, object: nil)

        var items = [UIBarButtonItem]()
        let editItem = UIBarButtonItem(barButtonSystemItem: .edit, target: self, action: #selector(editMessages(_:)))
        items.append(editItem)
        self.navigationItem.rightBarButtonItems = items

    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        if !returnFromSuspend {
            if contact == nil {
                let frame = self.view.bounds
                let viewRect = CGRect(x: 0.0, y: 0.0, width: frame.width * 0.8, height: frame.height * 0.7)
                selectView = SelectContactView(frame: viewRect)
                selectView!.contentView.backgroundColor = UIColor.init(gradientStyle: .topToBottom, withFrame: viewRect, andColors: [UIColor.flatPowderBlue, UIColor.flatSkyBlue])
                selectView!.contentView.alpha = 0.8
                selectView!.center = self.view.center
                self.view.addSubview(self.selectView!)
            }
            else {
                conversationDataSource = ConversationDataSource.init(collectionNode: asyncCollectionNode, contact: contact!)
                deferredDataSource.conversationDataSource = conversationDataSource!
                self.navigationItem.title = self.contact!.displayName
                self.contact!.conversation!.markMessagesRead()
                self.contact!.conversation!.isVisible = true
            }
        }

        NotificationCenter.default.addObserver(self, selector: #selector(appResumed(_:)),
                                               name: Notifications.AppResumed, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(appSuspended(_:)),
                                               name: Notifications.AppSuspended, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(presentAlert(_:)),
                                               name: Notifications.PresentAlert, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardDidShow(_:)),
                                               name: .UIKeyboardWillShow, object: nil)

    }

    override func viewWillDisappear(_ animated: Bool) {

        NotificationCenter.default.removeObserver(self, name: Notifications.AppResumed, object: nil)
        NotificationCenter.default.removeObserver(self, name: Notifications.AppSuspended, object: nil)
        NotificationCenter.default.removeObserver(self, name: Notifications.PresentAlert, object: nil)
        NotificationCenter.default.removeObserver(self, name: .UIKeyboardWillShow, object: nil)
        contact?.conversation!.isVisible = false

    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    override func didPressRightButton(_ sender: Any?) {
    
        let text = textView.text
        let textMessage = TextMessage(text: text!, contact: contact!)
        textMessage.timestamp = contact!.conversation!.getTimestamp()
        conversationDataSource!.appendMessage(textMessage)

        DispatchQueue.global().async {
            do {
                try self.contact!.conversation!.sendMessage(textMessage)
            }
            catch {
                let alertColor = UIColor.flatSand
                RKDropdownAlert.title("Message Error", message: "Failed to send message", backgroundColor: alertColor,
                                      textColor: ContrastColorOf(alertColor, returnFlat: true),
                                      time: 2, delegate: self)
            }
        }

        super.didPressRightButton(sender)

    }

    @objc func cancelEdit(_ item: Any) {
        
        var items = [UIBarButtonItem]()
        let editItem = UIBarButtonItem(barButtonSystemItem: .edit, target: self, action: #selector(editMessages(_:)))
        items.append(editItem)
        self.navigationItem.rightBarButtonItems = items
        
    }

    @objc func clearMessages(_ item: Any) {

        conversationDataSource!.clearMessages()
        contact!.conversation!.clearMessages()

        var items = [UIBarButtonItem]()
        let editItem = UIBarButtonItem(barButtonSystemItem: .edit, target: self, action: #selector(editMessages(_:)))
        items.append(editItem)
        self.navigationItem.rightBarButtonItems = items
        
    }
    
    @objc func editMessages(_ item: Any) {

        var items = [UIBarButtonItem]()
        let cancelItem = UIBarButtonItem(barButtonSystemItem: .cancel, target: self,
                                         action: #selector(cancelEdit(_:)))
        items.append(cancelItem)
        let clearItem = UIBarButtonItem(title: "Clear", style: .plain,
                                        target: self, action: #selector(clearMessages(_:)))
        items.append(clearItem)
        self.navigationItem.rightBarButtonItems = items

    }

    @objc func appResumed(_ notification: Notification) {

        if suspended {
            suspended = false
            returnFromSuspend = true
            if let info = notification.userInfo {
                let suspendedTime = info["suspendedTime"] as! NSNumber
                if suspendedTime.intValue > 0 && suspendedTime.intValue < 180 {
                    authView!.suspended = true
                }
            }
            else {
                let auth = Authenticator()
                auth.logout()
            }
            DispatchQueue.main.async {
                self.present(self.authView!, animated: true, completion: nil)
                self.view.alpha = 1.0
            }

        }

    }
    
    @objc func appSuspended(_ notification: Notification) {

        suspended = true
        DispatchQueue.main.async {
            self.view.alpha = 0.2
        }

    }
    
    @objc func keyboardDidShow(_ notification: Notification) {

        self.scrollCollectionViewToBottom()

    }

    @objc func contactSelected(_ notification: Notification) {

        contact = notification.object as? Contact
        conversationDataSource = ConversationDataSource.init(collectionNode: asyncCollectionNode, contact: contact!)
        deferredDataSource.conversationDataSource = conversationDataSource!
        DispatchQueue.main.async {
            self.navigationItem.title = self.contact!.displayName
            self.contact!.conversation!.isVisible = true
        }

    }

    @objc func messageAdded(_ notification: Notification) {

        DispatchQueue.main.async {
            self.scrollCollectionViewToBottom()
        }

    }

    @objc func presentAlert(_ notification: Notification) {
        
        let userInfo = notification.userInfo!
        let title = userInfo["title"] as? String
        let message = userInfo["message"] as? String
        DispatchQueue.main.async {
            let alertColor = UIColor.flatSand
            RKDropdownAlert.title(title, message: message, backgroundColor: alertColor,
                                  textColor: ContrastColorOf(alertColor, returnFlat: true),
                                  time: 2, delegate: self)
        }
        
    }
    
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
    }
    */

    func dropdownAlertWasTapped(_ alert: RKDropdownAlert!) -> Bool {
        return true
    }
    
    func dropdownAlertWasDismissed() -> Bool {
        return true
    }
    
}
